File: scripts/insight_generation.py
import openai
import os
import sys
import re

api_key = os.environ.get("OPENAI_API_KEY")
openai.api_key = api_key  # Set API key for the library

def generate_recommendations(cv_data, job_position):
    try:
        # Craft prompt
        
        prompt = f"Hello! I'm seeking advice on improving my CV for a {job_position} position. Below is the content of my CV:\n\n"
        # for section in cv_data:
        #     for key, value in section.items():
        #         prompt += f"CV section: {key}\nContent: {value}\n\n"
        for key, value in cv_data.items():
            prompt += f"CV section: {key}\nContent: {value}\n\n"
        prompt += "\nBased on this information and the targeted position, could you please provide recommendations on how I can enhance my CV to better align with the requirements and stand out to potential employers?"

        # Generate recommendations
        chat_completion = openai.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[
                {
                    "role": "system",
                    "content": prompt
                }
            ]
        )
        recommendations = chat_completion.choices[0].message.content.strip()

        # Parse recommendations for more specific details
        parsed_recommendations = parse_recommendations(recommendations)

        return parsed_recommendations
    except Exception as e:
        print(f"An error occurred: {e}")
        return ""

def parse_recommendations(recommendations):
    # Parse recommendations for more specific details
    
    return recommendations

if __name__ == "__main__":
    try:
        # cv_data = json.loads(sys.argv[1])
        data = sys.argv[1]

        data_string_trimmed = data[1:-1]


        # Splitting the string on commas not followed by space using regex
        data_parts = re.split(r',(?![\s])', data_string_trimmed)

        # Creating dictionary by splitting keys and values at colon
        data_dict = {part.split(':')[0]: ':'.join(part.split(':')[1:]) for part in data_parts}

        job_position = sys.argv[2]
        
        recommendations = generate_recommendations(data_dict, job_position)

        # Print recommendations
        print("Recommendations to improve your CV:")
        print(recommendations)
    except Exception as e:
        print(f"An error occurred: {e}")

File: scripts/insight_generation.js
const OpenAI = require('openai')

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
})

async function generate_recommendations(cvContent, jobPosition) {
  try {
    let prompt = `Hello! I'm seeking advice on improving my CV for a ${jobPosition} position. Below is the content of my CV:\n\n${cvContent}\n\n`
    prompt += `Based on this information and the targeted position, could you please provide recommendations on how I can enhance my CV to better align with the requirements and stand out to potential employers?`

    const response = await openai.chat.completions.create({
      model: 'gpt-3.5-turbo-0125',
      messages: [{ role: 'user', content: prompt }],
    })

    const recommendationsString = response.choices[0].message.content.trim()
    if (recommendationsString.length < 10) {
      return [
        {
          title: 'No Recommendations',
          main_content:
            'No specific recommendations could be generated based on the provided CV content.',
        },
      ]
    }

    // Split the recommendations into an array based on some delimiter
    const recommendationsArray = recommendationsString
      .split('\n')
      .filter(Boolean)

    // Map the array to recommendation objects containing the title without numbering and main content
    const recommendations = recommendationsArray.map((recommendation) => {
      const parts = recommendation.split(':')
      let title = parts[0].trim()
      // Remove any numbering or "Recommendation" indication from the title
      title = title.replace(/^\d+\.\s*|\s*Recommendation\s*/i, '')
      const main_content = parts.slice(1).join(':').trim()
      return {
        title: title || 'General Recommendation',
        main_content: main_content,
      }
    })

    return recommendations
  } catch (e) {
    console.log(`An error occurred: ${e}`)
    return [
      {
        title: 'Error',
        main_content: 'An error occurred while generating recommendations.',
      },
    ]
  }
}

module.exports = generate_recommendations
